spec/highlighting.spec.js

```diff
@@ -54,4 +53,86 @@ describe('Highlighting', function () {
     })
   })
 
+  describe('highlightSupport', () => {
+    beforeEach(() => {
+      this.$div = $('<div>Foobarista</div>').appendTo(document.body)
+      this.editable = new Editable()
+      this.editable.add(this.$div)
+      const range = rangy.createRange()
+      range.selectNodeContents(this.$div[0])
+      this.selection = new Selection(this.$div[0], range)
+    })
+
+    afterEach(() => {
+      this.$div.remove()
+      this.editable.off()
+      this.editable = undefined
+    })
+
+    it('can highlight text range matches', () => {
+      const startIndex = this.editable.highlight({
+        editableHost: this.$div[0],
+        text: 'bari',
+        highlightId: 'myId',
+        textRange: { start: 3, end: 7 }
+      })
+      const highlightSpan = this.$div.find('[data-word-id="myId"]')
+      expect(highlightSpan.length).toEqual(1)
+      expect(highlightSpan.text()).toEqual('bari')
+      expect(startIndex).toEqual(3)
+    })
+
+    it('can extract a text-based range object', () => {
+      const editableHost = this.$div[0]
+      this.editable.highlight({
+        editableHost,
+        text: 'bari',
+        highlightId: 'myId',
+        textRange: { start: 3, end: 7 }
+      })
+
+      this.editable.highlight({
+        editableHost,
+        text: 'Fo',
+        highlightId: 'mySecondId',
+        textRange: { start: 0, end: 2 }
+      })
+      const extractedHighlightPositions = highlightSupport.extractHighlightedRanges(editableHost, this.selection)
+      expect(extractedHighlightPositions.myId).toBeDefined()
+      expect(extractedHighlightPositions.myId.start).toEqual(3)
+      expect(extractedHighlightPositions.myId.end).toEqual(7)
+      expect(extractedHighlightPositions.mySecondId).toBeDefined()
+      expect(extractedHighlightPositions.mySecondId.start).toEqual(0)
+      expect(extractedHighlightPositions.mySecondId.end).toEqual(2)
+    })
+
+    it('skips if a highlight with the given id is already present', () => {
+      this.editable.highlight({
+        editableHost: this.$div[0],
+        text: 'bari',
+        highlightId: 'myId',
+        textRange: { start: 3, end: 7 }
+      })
+      this.editable.highlight({
+        editableHost: this.$div[0],
+        text: 'Foobari',
+        highlightId: 'myId',
+        textRange: { start: 0, end: 7 }
+      })
+      const highlightSpan = this.$div.find('[data-word-id="myId"]')
+      expect(highlightSpan.length).toEqual(1)
+      expect(highlightSpan.text()).toEqual('bari')
+    })
+
+    it('skips if an invalid range object was passed', () => {
+      this.editable.highlight({
+        editableHost: this.$div[0],
+        text: 'bari',
+        highlightId: 'myId',
+        textRange: { foo: 3, bar: 7 }
+      })
+      const highlightSpan = this.$div.find('[data-word-id="myId"]')
+      expect(highlightSpan.length).toEqual(0)
+    })
+  })
```

Comment:
- minimalistic ranges
- 0-1 positions
- containing / overlapping ranges
- succeeding ranges

Reply:
more use cases:
end position
ranges with newlines within
two identical ranges
markup (<b> and all possible contains and intersections)
other highlights overlapps

Reply:
- textrange export on selection